lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs

```diff
@@ -93,54 +93,81 @@ public Thread StartTaskOnNewThread()
                     Logger.Log(Level.Info, "Task Call Finished");
-                    _currentStatus.SetResult(result);
 
-                    const Level resultLogLevel = Level.Verbose;
+                    // Run the handlers for `TaskStop`
+                    _currentStatus.RunTaskStopHandlers();
 
-                    if (Logger.IsLoggable(resultLogLevel) && result != null && result.Length > 0)
+                    // Log the result
+                    const Level resultLogLevel = Level.Verbose;
+                    if (Logger.IsLoggable(resultLogLevel) && resultReturnedByTask != null && resultReturnedByTask.Length > 0)
                     {
                         Logger.Log(resultLogLevel,
-                            "Task running result:\r\n" + System.Text.Encoding.Default.GetString(result));
+                            "Task running result:\r\n" + System.Text.Encoding.Default.GetString(resultReturnedByTask));
                     }
                 }
                 catch (TaskStartHandlerException e)
                 {
                     Logger.Log(Level.Info, "TaskRuntime::TaskStartHandlerException");
                     _currentStatus.SetException(e.InnerException);
+                    returnResultToDriver = false;
                 }
                 catch (TaskStopHandlerException e)
                 {
                     Logger.Log(Level.Info, "TaskRuntime::TaskStopHandlerException");
                     _currentStatus.SetException(e.InnerException);
+                    returnResultToDriver = false;
                 }
                 catch (Exception e)
                 {
                     Logger.Log(Level.Info, "TaskRuntime::Exception {0}", e.GetType());
                     _currentStatus.SetException(e);
+                    returnResultToDriver = false;
                 }
                 finally
                 {
                     try
                     {
-                        if (_userTask != null)
-                        {
-                            _userTask.Dispose();
-                        }
+                        _userTask.Dispose();
                     }
                     catch (Exception e)
                     {
-                        var msg = "Exception during Task Dispose in task Call()";
-                        Logger.Log(Level.Error, msg);
-                        throw new InvalidOperationException(msg, e);
+                        exceptionThrownByTaskDispose = new InvalidOperationException("Exception during Task Dispose in task Call()", e);
                     }
                 }
+
+                // Inform the driver about the result.
+                if (returnResultToDriver)
+                {   
+                    _currentStatus.SetResult(resultReturnedByTask);
+                }
+
+                // If the ITask.Dispose() method threw an Exception, crash the Evaluator.
+                if (exceptionThrownByTaskDispose != null)
```

Comment:
So if Dispose throws exception and task does get result, it still sets result first, then  throw exception to crash the evaluator?